src/twisted/internet/_sslverify.py

```diff
@@ -1682,22 +1646,40 @@ def _makeContext(self):
 
         if self.dhParameters:
             ctx.load_tmp_dh(self.dhParameters._dhFile.path)
-        ctx.set_cipher_list(self._cipherString.encode("ascii"))
 
         self._ecChooser.configureECDHCurve(ctx)
+        if self._contextForServerName is not None:
+            ctx.set_tlsext_servername_callback(
+                contextFactoryToServernameCallback(self._contextForServerName)
+            )
+        setupForALPN(ctx, self._acceptableProtocols or ())
+        if not skipCiphers:
+            ctx.set_cipher_list(self._cipherString.encode("ascii"))
+        return ctx
 
-        if self._acceptableProtocols:
-            # Try to set NPN and ALPN. _acceptableProtocols cannot be set by
-            # the constructor unless at least one mechanism is supported.
-            _setAcceptableProtocols(ctx, self._acceptableProtocols)
 
-        return ctx
+def contextFactoryToServernameCallback(
```

Comment:
Is this finalized ? I see the TODO comment.
 
Maybe renamed it to `_contextFactoryToServernameCallback` to signal that this is never exposed to end users.

It's a bit hard to track what is really private and what is exposed/aliased or returned from other public APIs.